Automatically set the ajax_load request parameter.

thet opened the pull request:
Related:
https://github.com/plone/Products.CMFPlone/pull/4188
https://github.com/plone/plone.base/pull/85
https://github.com/plone/plonetheme.barceloneta/pull/404
https://github.com/plone/plone.app.portlets/pull/203


Plone 6.1: https://github.com/plone/Products.CMFPlone/pull/4188  (this one)
Plone 6.2: https://github.com/plone/Products.CMFPlone/pull/4169

Zope maintains the "HTTP_X_REQUESTED_WITH" request header. If this is set to "XMLHttpRequest", we have an AJAX request. In this case the ajax_load parameter is set to 1 directly on the request object.

This should make use for the ajax_main_template for any AJAX request, regardless if the ajax_load parameter was manually set or not and potentially speed up Plone Classic UI rendering.

mister-roboto commented:
@thet thanks for creating this Pull Request and helping to improve Plone!

TL;DR: Finish pushing changes, pass all other checks, then paste a comment:
```
@jenkins-plone-org please run jobs
```

To ensure that these changes do not break other parts of Plone, the Plone test suite matrix needs to pass, but it takes 30-60 min.  Other CI checks are usually much faster and the Plone Jenkins resources are limited, so when done pushing changes and all other checks pass either [start all Jenkins PR jobs yourself](https://jenkinsploneorg.readthedocs.io/en/latest/run-pull-request-jobs.html#run-a-pull-request-job), or simply add the comment above in this PR to start all the jobs automatically.

Happy hacking!

thet commented:
@jenkins-plone-org please run jobs

ale-rt commented on Products/CMFPlone/browser/templates/ajax_main_template.pt:
```diff
@@ -16,7 +16,7 @@
           checkPermission python:context.restrictedTraverse('portal_membership').checkPermission;
           ajax_include_head python:request.get('ajax_include_head', False);
-          ajax_load python:False;"
+          ajax_load python:True;"
```
I have no idea why this `ajax_main_template` has `ajax_load=False` in the first place. This variable seems really obscure defined like this.

ale-rt commented on Products/CMFPlone/traversal.py:
```diff
@@ -72,3 +75,17 @@ def get_zope_page_template_engine(engine):
+
+    # If ajax_load was already set to a true-ish or false-ish value, don't set
+    # it again.
+    if (
+        request.getHeader("HTTP_X_REQUESTED_WITH") == "XMLHttpRequest"
+        and not request.get("ajax_load", MARKER) is not MARKER
+    ):
+        # Directly set on the request object
+        request.set("ajax_load", True)
```
I do not like this approach.
You are checking the request, fiddling with the request parameters, and then checking the request again.

IMO this is too complicated.
There should be some come somewhere that has a method `is_ajax_load` (or whatever other name) that is the unique source of truth about the fact that we are loading stuff through AJAX.

Ideally, this should be a method in a view (`plone_context_state`?) so that it is easily configurable (e.g. you might want to always load some portal type as if it was called as AJAX, add a check for the referrer or whatever).

BTW even if you want to have this traverser, the second condition has to be simplified because it contains a double negation.

Something like:
```python
request.get("ajax_load", MARKER) is MARKER
```
or even:
```python
"ajax_load" not in request
```
is preferable.

thet replied:
LOL, `and not request.get("ajax_load", MARKER) is not MARKER` .... yes, i'll take your last suggestion - this is much better :dagger:

thet commented:
@ale-rt I have addressed your comments in https://github.com/plone/Products.CMFPlone/pull/4169

I agree with @mauritsvanrees this should probably not be merged into Plone 6.1 but 6.2 instead.
